[CAS] swift dependency scanning using CAS for compiler caching

cachemeifyoucan opened the pull request:
Teach swift dependency scanner to scan dependency with a CAS for compiler caching.

cachemeifyoucan committed:
[CAS] Swift CAS dependency scanning for explicit build

cachemeifyoucan committed:
[CAS] Support clang include tree

Using clang include tree for dependency scanning and building for all
clang modules and PCHs.

artemcm commented on include/swift/Frontend/ModuleInterfaceLoader.h:
```diff
@@ -201,18 +260,22 @@ struct ExplicitSwiftModuleInputInfo {
   bool isFramework = false;
   // A flag that indicates whether this module is a system module
   bool isSystem = false;
+  // The cache key for clang module.
```
Typo, I think, this is for a Swift module.

artemcm commented on include/swift/Serialization/ModuleDependencyScanner.h:
```diff
@@ -41,18 +42,22 @@ namespace swift {
                               Identifier moduleName,
                               InterfaceSubContextDelegate &astDelegate,
-                              ScannerKind kind = MDS_plain)
+                              ScannerKind kind = MDS_plain,
```
Hmm, not a comment on this PR at all, but I ought to go rip this out altogether...

artemcm commented on lib/AST/ModuleDependencies.cpp:
```diff
@@ -253,25 +382,96 @@ void ModuleDependencyInfo::addBridgingModuleDependency(
+SwiftDependencyScanningService::SwiftDependencyScanningService() {
+  ClangScanningService.emplace(
+      clang::tooling::dependencies::ScanningMode::DependencyDirectivesScan,
+      clang::tooling::dependencies::ScanningOutputFormat::FullTree,
```
Is this to be set irrespective of whether CAS is used?

artemcm commented on lib/AST/ModuleDependencies.cpp:
```diff
@@ -253,25 +382,96 @@ void ModuleDependencyInfo::addBridgingModuleDependency(
+  CommonDependencyFiles.emplace_back(SDKSettingPath.data(),
+                                     SDKSettingPath.size());
+
+  // Add Legacy layout file (maybe just hard code instead of searching).
```
What's the context here, and why are we only using the first path?

artemcm commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -175,18 +181,89 @@ void ClangImporter::recordModuleDependencies(
+        swiftArgs.push_back(overlay);
+      }
+    } else {
+      // HACK: find the -ivfsoverlay option from clang scanner and pass to
```
Could you please explain the reasoning in a comment? How/Why could there be overlays here not covered by the above `if` conditional?

cachemeifyoucan replied:
I forgot about this hack completely. This is rdar://105684525

artemcm commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -175,18 +181,89 @@ void ClangImporter::recordModuleDependencies(
     // Add args reported by the scanner.
-    llvm::for_each(clangModuleDep.BuildArguments, addClangArg);
+
+    // Round-trip clang args to canonicalize and clear the options that swift
```
Nit: What do you think of factoring out the code that does this from the overall flow of this method?

cachemeifyoucan replied:
I think this should go the other direction. All operations on clang arguments should be done this way. Maybe I will do a cleanup for all the string arg manipulation in swift to use this round trip instead.

artemcm commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -200,6 +277,85 @@ void ClangImporter::recordModuleDependencies(
+  swiftArgs.push_back("-frontend");
+
+  // Swift frontend action: -emit-pcm
+  swiftArgs.push_back("-emit-pch");
```
We need to consider carefully what is added here versus what is added in the driver's `generatePCHJob`, which also includes `addCommonFrontendOptions`, which may have additional context on top of what we know here.

cachemeifyoucan replied:
I can post the patch for swift driver too. I haven't got time to write extensive tests for that.

cachemeifyoucan commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -172,6 +184,10 @@ SWIFTSCAN_PUBLIC bool
     swiftscan_module_details_t details);
 
+SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+swiftscan_swift_binary_detail_get_module_cache_key(
```
`cas_fs_root_id` and `include_tree_root` capture the file dependencies (and more) for the compilation. For a binary deps, there is no compilation job, thus it is not needed.

cachemeifyoucan commented on include/swift/AST/ModuleDependencies.h:
```diff
@@ -141,6 +155,19 @@ struct CommonSwiftTextualModuleDependencyDetails {
+  std::vector<std::string> buildCommandLine;
+
+  /// The Swift frontend invocation arguments to build bridging header.
+  std::vector<std::string> bridgingHeaderBuildCommandLine;
```
Sure. It can be moved.

cachemeifyoucan commented on include/swift/AST/ModuleDependencies.h:
```diff
@@ -297,20 +348,35 @@ class ClangModuleDependencyStorage : public ModuleDependencyInfoStorageBase {
+  std::string clangIncludeTreeRoot;
+
+  /// The cache key for the produced module.
+  std::string moduleCacheKey;
```
Place holder currently doesn't have one because it is not supported, but moving to base class makes sense.

artemcm replied:
Yeah, makes sense. I'm looking to try and remove the notion of placeholders soon.

cachemeifyoucan commented on include/swift/AST/ModuleDependencies.h:
```diff
@@ -141,6 +155,19 @@ struct CommonSwiftTextualModuleDependencyDetails {
+
+  /// The Swift frontend invocation arguments to build the Swift module from the
+  /// interface.
+  std::vector<std::string> buildCommandLine;
```
Currently, it is used as extra swift command that needs to be passed, which specifies some caching specific options that constructed from due-scanning. I am currently working on a patch for dep scanner to also return full cc1 command for source module, and I wonder if I should add another list of cc1 command, or just append those here too.

artemcm replied:
I think having it as a separate list would be nice with minimal downside. Then the client (driver) could be very specific with what it itself is passing, what it is using from the scanner, and what it is passing for underlying Clang compilation. I'm wary of making sure we have well-defined categories of flags coming from all possible sources, as to minimize confusion for how compilation commands are formulated.

cachemeifyoucan committed:
Address review feedback

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan committed:
[ScanDependency] Fix dependency output json format

Remove a trailing common that makes the JSON output from
`-scan-dependencies` to be invalide JSON format.

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please smoke test

cachemeifyoucan commented:
@swift-ci please test windows platform
